core/DeviceDetector/DeviceDetectorFactory.php

```diff
@@ -20,37 +20,39 @@ class DeviceDetectorFactory
      * Returns an instance of DeviceDetector for the given user agent. Uses template method pattern
      * and calls getDeviceDetectionInfo() when it doesn't find a matching instance in the cache.
      * @param string $userAgent
+     * @param array $clientHints
      * @return DeviceDetector|mixed
      */
-    public function makeInstance($userAgent)
+    public function makeInstance($userAgent, $clientHints)
     {
-        $userAgent = self::getNormalizedUserAgent($userAgent);
+        $cacheKey = self::getNormalizedUserAgent($userAgent, $clientHints);
 
-        if (array_key_exists($userAgent, self::$deviceDetectorInstances)) {
-            return self::$deviceDetectorInstances[$userAgent];
+        if (array_key_exists($cacheKey, self::$deviceDetectorInstances)) {
+            return self::$deviceDetectorInstances[$cacheKey];
         }
 
-        $deviceDetector = $this->getDeviceDetectionInfo($userAgent);
+        $deviceDetector = $this->getDeviceDetectionInfo($userAgent, $clientHints);
 
-        self::$deviceDetectorInstances[$userAgent] = $deviceDetector;
+        self::$deviceDetectorInstances[$cacheKey] = $deviceDetector;
 
         return $deviceDetector;
     }
 
-    public static function getNormalizedUserAgent($userAgent)
+    public static function getNormalizedUserAgent($userAgent, $clientHints = [])
     {
-        return mb_substr(trim($userAgent), 0, 500);
+        return mb_substr(md5(json_encode($clientHints) ?? '') . trim($userAgent), 0, 500);
```

Comment:
I don't know what this is used for exactly, but keep in mind that user agents are encouraged to randomize the exact values, so this might not be useful as a cache string. (See https://wicg.github.io/ua-client-hints/#grease)

Reply:
sort it alphabetically first might work?